File: sidecar/reminders_sidecar/sync.py
"""
Sync engine: pulls iCloud into the cache and pushes local edits back out.

Two pull modes:

  full_sync   -- every list, every reminder. Used on first run and on demand.
                 The real account this was built against holds ~2,900 reminders
                 across 14 lists, so this reports progress rather than blocking
                 silently.

  delta_sync  -- iter_changes(since=cursor). Cheap, runs on the background
                 timer. It only ever reports reminders, so lists are refreshed
                 wholesale on every delta pass too. That is not redundancy; it
                 is the only way a renamed or deleted list is ever noticed.

Push is an outbox: local edits land in the cache immediately, then drain to
iCloud. A push whose base change tag no longer matches the server is recorded
as a conflict instead of overwriting -- the phone writes to the same records
and was observed replacing fields wholesale rather than merging.
"""

from __future__ import annotations

import logging
import threading
from datetime import datetime, timezone
from typing import Callable, Optional

from .db import Cache, to_iso, utcnow
from .icloud import AuthRequired, ConflictError, ICloudClient, SidecarError

LOGGER = logging.getLogger(__name__)

CURSOR_KEY = "sync_cursor"
LAST_FULL_KEY = "last_full_sync"
LAST_SYNC_KEY = "last_sync"


class SyncEngine:
    def __init__(
        self,
        cache: Cache,
        client: ICloudClient,
        emit: Optional[Callable[[str, dict], None]] = None,
    ):
        self.cache = cache
        self.client = client
        self.emit = emit or (lambda _e, _d: None)
        self._lock = threading.Lock()
        # Writes kick a push each, so several drains can be requested at once.
        # Only one may run: two draining the same rows would push twice.
        self._push_lock = threading.Lock()
        self._running = False

    @property
    def running(self) -> bool:
        return self._running

    # ------------------------------------------------------------------ pull
    def full_sync(self) -> dict:
        """Rebuild the cache from scratch. Safe to run at any time."""
        with self._lock:
            self._running = True
            try:
                self.emit("sync_started", {"mode": "full", "determinate": True})

                # Take the cursor BEFORE reading, so anything that changes while
                # we page through is picked up by the next delta rather than lost.
                cursor = self.client.sync_cursor()

                lists = self.client.lists()
                self.cache.replace_lists(lists)

                # Progress is weighted by each list's reminder count rather than
                # counting lists, because the work is wildly uneven -- one list
                # on the account this was built against holds 1,219 of the 2,900
                # records. A bar that moves in equal steps per list would sit at
                # 90% for most of the sync.
                work = [l for l in lists if not l.get("is_group")]
                weights = [max(int(l.get("count") or 0), 1) for l in work]
                budget = sum(weights) or 1
                self.emit(
                    "sync_progress",
                    {
                        "stage": "lists",
                        "count": len(lists),
                        "expected": sum(int(l.get("count") or 0) for l in work),
                        "percent": 0.0,
                        "done": 0,
                        "of": len(work),
                    },
                )

                total = 0
                spent = 0
                for i, l in enumerate(work):
                    reminders, tags_by = self.client.reminders_for(l["id"])
                    self.cache.upsert_reminders(reminders)
                    for rid, tags in tags_by.items():
                        self.cache.replace_tags_for(rid, tags)
                    total += len(reminders)
                    spent += weights[i]
                    self.emit(
                        "sync_progress",
                        {
                            "stage": "reminders",
                            "list": l["title"],
                            "index": i + 1,
                            "done": i + 1,
                            "of": len(work),
                            "total": total,
                            "percent": round(100.0 * spent / budget, 1),
                        },
                    )

                if cursor:
                    self.cache.set_meta(CURSOR_KEY, cursor)
                now = to_iso(utcnow())
                self.cache.set_meta(LAST_FULL_KEY, now)
                self.cache.set_meta(LAST_SYNC_KEY, now)
                self.emit("sync_finished", {"mode": "full", "reminders": total})
                return {"mode": "full", "lists": len(lists), "reminders": total}
            finally:
                self._running = False

    def delta_sync(self) -> dict:
        """Apply changes since the stored cursor. Falls back to full if unset."""
        cursor = self.cache.get_meta(CURSOR_KEY)
        if not cursor:
            return self.full_sync()

        with self._lock:
            self._running = True
            try:
                # A delta has no measurable size until the changes come back, so
                # the bar animates rather than lying about a percentage.
                self.emit("sync_started", {"mode": "delta", "determinate": False})

                # Lists never appear in the delta stream -- always refresh them.
                lists = self.client.lists()
                self.cache.replace_lists(lists)
                self.emit("sync_progress", {"stage": "lists", "count": len(lists)})

                changes = self.client.changes_since(cursor)
                updated = [c for c in changes if not c.get("deleted")]
                deleted = [c for c in changes if c.get("deleted")]
                self.emit(
                    "sync_progress",
                    {"stage": "changes", "total": len(changes)},
                )

                if updated:
                    self.cache.upsert_reminders(updated)
                    self._refresh_tags_for_changed(updated)
                for d in deleted:
                    self.cache.apply_local_edit(d["id"], {"deleted": 1})
                    self.cache.clear_dirty(d["id"], None)

                new_cursor = self.client.sync_cursor()
                if new_cursor:
                    self.cache.set_meta(CURSOR_KEY, new_cursor)
                self.cache.set_meta(LAST_SYNC_KEY, to_iso(utcnow()))

                self.emit(
                    "sync_finished",
                    {"mode": "delta", "updated": len(updated), "deleted": len(deleted)},
                )
                return {
                    "mode": "delta",
                    "updated": len(updated),
                    "deleted": len(deleted),
                }
            finally:
                self._running = False

    def _refresh_tags_for_changed(self, changed: list[dict]) -> None:
        """
        Re-read tags for reminders the delta touched.

        Hashtag records don't come through iter_changes, and the phone rewrites
        HashtagIDs wholesale, so a changed reminder's tags cannot be trusted to
        be what we already had. Only reminders that actually reference hashtags
        are re-read, to keep this cheap.
        """
        for c in changed:
            ids = c.get("hashtag_ids")
            if ids is None:
                continue
            if not ids:
                self.cache.replace_tags_for(c["id"], [])
                continue
            try:
                rem = self.client._service().get(c["id"])  # noqa: SLF001
                self.cache.replace_tags_for(c["id"], self.client.tags_for(rem))
            except Exception as exc:  # noqa: BLE001 - tags are cosmetic; never fail sync
                LOGGER.debug("tag refresh failed for %s: %s", c["id"], exc)

    # ------------------------------------------------------------------ push
    def flush_outbox(self) -> dict:
        """
        Drain queued local edits to iCloud.

        Conflicts are recorded, not resolved: the server copy wins in the cache
        so the UI matches reality, and the local version is preserved in the
        conflicts table for the user to re-apply.
        """
        if not self._push_lock.acquire(blocking=False):
            # Another drain is already in flight; it will pick up whatever was
            # queued in the meantime, so this call is redundant rather than lost.
            return {"pushed": 0, "conflicts": 0, "failed": 0, "skipped": True}
        try:
            return self._drain()
        finally:
            self._push_lock.release()

    def push_now(self) -> dict:
        """A push on its own, with the same expired-session recovery as a sync."""
        return self._with_reauth(self.flush_outbox)

    def _drain(self) -> dict:
        pushed = 0
        conflicts = 0
        failed = 0
        expired: Optional[SidecarError] = None

        for item in self.cache.pending():
            seq = item["seq"]
            op = item["op"]
            rid = item["reminder_id"]
            payload = item["payload"]

            try:
                if op == "create":
                    created = self.client.create(payload)
                    self.cache.replace_id(rid, created["id"])
                    self.cache.upsert_reminders([created])
                    self.cache.clear_dirty(created["id"], created.get("change_tag"))
                elif op == "update":
                    updated = self.client.update(rid, payload, item.get("base_tag"))
                    self.cache.clear_dirty(rid, updated.get("change_tag"))
                    self.cache.upsert_reminders([updated])
                elif op == "delete":
                    self.client.delete(rid)
                    self.cache.clear_dirty(rid, None)
                else:
                    LOGGER.warning("unknown outbox op %r", op)
                self.cache.dequeue(seq)
                pushed += 1

            except ConflictError as exc:
                local = self.cache.reminder(rid) or {}
                remote = {}
                if exc.detail:
                    try:
                        import json

                        remote = json.loads(exc.detail)
                    except ValueError:
                        remote = {"raw": exc.detail}
                self.cache.record_conflict(rid, local, remote)
                # Server truth into the cache; the local edit lives on in the
                # conflict row rather than being silently dropped or forced.
                if remote:
                    self.cache.clear_dirty(rid, remote.get("change_tag"))
                    self.cache.upsert_reminders([remote])
                self.cache.dequeue(seq)
                conflicts += 1
                self.emit("conflict", {"reminder_id": rid})

            except SidecarError as exc:
                self.cache.record_failure(seq, exc.message)
                failed += 1
                self.emit("push_failed", {"reminder_id": rid, "error": exc.to_dict()})
                # Auth problems will fail every remaining item too; stop early.
                if exc.code in ("AUTH_REQUIRED", "2FA_REQUIRED", "TERMS_REQUIRED"):
                    # An expired session is recoverable, so it has to escape this
                    # loop rather than being absorbed into a failure count --
                    # otherwise the caller never learns there is anything to
                    # retry. The item stays queued; only its attempt count moved.
                    if exc.code == "AUTH_REQUIRED":
                        expired = exc
                    break

        if expired is not None:
            raise expired
        return {"pushed": pushed, "conflicts": conflicts, "failed": failed}

    def sync_now(self, full: bool = False) -> dict:
        """Push first, then pull, so local edits aren't clobbered by our own pull."""
        push = self._with_reauth(self.flush_outbox)
        pull = self._with_reauth(self.full_sync if full else self.delta_sync)
        return {"push": push, "pull": pull}

    def _with_reauth(self, fn: Callable[[], dict]) -> dict:
        """
        Run a sync step, rebuilding the session once if Apple has expired it.

        iCloud session tokens do not last forever, and the old behaviour was to
        surface that as "sign in again" on a timer -- every background pass after
        expiry threw the user back to the login screen. A restore uses the saved
        credential and the trust token, so it needs nothing from them. Only one
        retry: if the second attempt still fails, the session really is gone and
        the sign-in prompt is the honest answer.
        """
        try:
            return fn()
        except AuthRequired:
            if not self.client.restore():
                raise
            LOGGER.info("iCloud session restored; retrying %s", getattr(fn, "__name__", fn))
            self.emit("auth_changed", self.client.status())
            return fn()

File: sidecar/src/sync.rs
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};

use chrono::Utc;
use serde_json::{Value, json};
use tokio::sync::{Mutex, mpsc::UnboundedSender};

use crate::cache::Cache;
use crate::error::{AppError, Result};
use crate::icloud::ICloudClient;
use crate::model::Reminder;
use crate::timeutil::to_iso;

pub const CURSOR_KEY: &str = "sync_cursor";
pub const LAST_FULL_KEY: &str = "last_full_sync";
pub const LAST_SYNC_KEY: &str = "last_sync";
const READ_PROTOCOL_KEY: &str = "cloudkit_read_protocol";
const READ_PROTOCOL_VERSION: &str = "2";

pub struct SyncEngine {
    cache: Arc<Cache>,
    client: Arc<Mutex<ICloudClient>>,
    events: UnboundedSender<Value>,
    running: AtomicBool,
    pushing: Mutex<()>,
}

struct RunningGuard<'a>(&'a AtomicBool);
impl Drop for RunningGuard<'_> {
    fn drop(&mut self) {
        self.0.store(false, Ordering::Release);
    }
}

impl SyncEngine {
    pub fn new(
        cache: Arc<Cache>,
        client: Arc<Mutex<ICloudClient>>,
        events: UnboundedSender<Value>,
    ) -> Self {
        Self {
            cache,
            client,
            events,
            running: AtomicBool::new(false),
            pushing: Mutex::new(()),
        }
    }
    pub fn running(&self) -> bool {
        self.running.load(Ordering::Acquire)
    }
    fn emit(&self, event: &str, data: Value) {
        let _ = self.events.send(json!({"event":event,"data":data}));
    }

    pub async fn sync_now(&self, full: bool) -> Result<Value> {
        if self.running.swap(true, Ordering::AcqRel) {
            return Ok(json!({"queued":false,"reason":"already running"}));
        }
        let _guard = RunningGuard(&self.running);
        let push = self.push_with_reauth().await?;
        let pull = self.pull_with_reauth(full).await?;
        Ok(json!({"push":push,"pull":pull}))
    }

    pub async fn push_now(&self) -> Result<Value> {
        self.push_with_reauth().await
    }

    /// Rebuild the session once after Apple refuses a request, or explain why not.
    ///
    /// The old version treated every failure to get back in as an expiry, and
    /// announced it as one. That is what "it signs me out all the time" was: the
    /// sticky sign-in notice went up over a dropped connection, and the only
    /// thing that clears that notice is signing in -- so a ten-second blip cost
    /// a password.
    async fn recover_session(&self) -> Result<()> {
        let outcome = {
            let mut client = self.client.lock().await;
            if client.awaiting_2fa() {
                // Someone is at the code box. Rebuilding from here hands Apple a
                // new challenge and retires the code they are halfway through
                // typing, which is how a correct code ends up rejected.
                return Err(AppError::TwoFactorRequired {
                    message: "Enter the verification code to finish signing in.".into(),
                    detail: String::new(),
                });
            }
            client.invalidate();
            let restored = client.restore().await;
            (restored, client.restore_is_retryable(), client.restore_detail())
        };
        let (restored, retryable, detail) = outcome;
        if restored {
            self.emit("auth_changed", self.client.lock().await.status());
            return Ok(());
        }
        if retryable {
            // Nothing here established that the session is gone, so do not say
            // that it is. The next pass tries again.
            return Err(AppError::Network {
                message: "Couldn't reach iCloud. Sync will try again shortly.".into(),
                detail,
            });
        }
        self.emit("auth_changed", self.client.lock().await.status());
        Err(AppError::AuthRequired {
            message: "Your iCloud session expired. Please sign in again.".into(),
            detail: String::new(),
        })
    }

    async fn push_with_reauth(&self) -> Result<Value> {
        match self.flush_outbox().await {
            Err(AppError::AuthRequired { .. }) => {
                self.recover_session().await?;
                self.flush_outbox().await
            }
            result => result,
        }
    }

    async fn pull_with_reauth(&self, full: bool) -> Result<Value> {
        // Protocol v1 could persist a valid cursor after Apple's unsupported
        // raw record queries returned an empty snapshot. Force one full pull
        // after upgrading so that stale empty state cannot survive the fix.
        let needs_full = full
            || self.cache.get_meta(CURSOR_KEY)?.is_none()
            || self.cache.get_meta(READ_PROTOCOL_KEY)?.as_deref()
                != Some(READ_PROTOCOL_VERSION);
        let first = if needs_full {
            self.full_sync().await
        } else {
            self.delta_sync().await
        };
        match first {
            Err(AppError::AuthRequired { .. }) => {
                self.recover_session().await?;
                if needs_full {
                    self.full_sync().await
                } else {
                    self.delta_sync().await
                }
            }
            result => result,
        }
    }

    async fn full_sync(&self) -> Result<Value> {
        self.emit("sync_started", json!({"mode":"full","determinate":true}));
        let cursor = self.client.lock().await.sync_cursor().await?;
        let lists = self.client.lock().await.lists().await?;
        self.cache.replace_lists(&lists)?;
        let work: Vec<_> = lists.iter().filter(|list| !list.is_group).collect();
        let weights: Vec<i64> = work.iter().map(|list| list.count.max(1)).collect();
        let budget = weights.iter().sum::<i64>().max(1);
        self.emit("sync_progress",json!({"stage":"lists","count":lists.len(),"expected":work.iter().map(|l|l.count).sum::<i64>(),"percent":0.0,"done":0,"of":work.len()}));
        let mut total = 0usize;
        let mut spent = 0i64;
        let mut reminder_ids = Vec::new();
        for (index, list) in work.iter().enumerate() {
            let reminders = self.client.lock().await.reminders_for(&list.id).await?;
            self.cache.upsert_reminders(&reminders)?;
            reminder_ids.extend(reminders.iter().map(|reminder| reminder.id.clone()));
            total += reminders.len();
            spent += weights[index];
            self.emit("sync_progress",json!({"stage":"reminders","list":list.title,"index":index+1,"done":index+1,"of":work.len(),"total":total,"percent":((1000*spent/budget) as f64)/10.0}));
        }
        let tags = self
            .client
            .lock()
            .await
            .tags_for_reminders(&reminder_ids)
            .await?;
        for id in &reminder_ids {
            self.cache
                .replace_tags_for(id, tags.get(id).map(Vec::as_slice).unwrap_or(&[]))?;
        }
        if let Some(value) = cursor.as_deref() {
            self.cache.set_meta(CURSOR_KEY, Some(value))?;
        }
        let now = to_iso(Utc::now());
        self.cache.set_meta(LAST_FULL_KEY, Some(&now))?;
        self.cache.set_meta(LAST_SYNC_KEY, Some(&now))?;
        self.cache
            .set_meta(READ_PROTOCOL_KEY, Some(READ_PROTOCOL_VERSION))?;
        self.emit("sync_finished", json!({"mode":"full","reminders":total}));
        Ok(json!({"mode":"full","lists":lists.len(),"reminders":total}))
    }

    async fn delta_sync(&self) -> Result<Value> {
        let cursor = self
            .cache
            .get_meta(CURSOR_KEY)?
            .ok_or_else(|| AppError::internal("Delta sync started without a cursor", ""))?;
        self.emit("sync_started", json!({"mode":"delta","determinate":false}));
        let lists = self.client.lock().await.lists().await?;
        self.cache.replace_lists(&lists)?;
        self.emit(
            "sync_progress",
            json!({"stage":"lists","count":lists.len()}),
        );
        let (updated, deleted, new_cursor) = self
            .client
            .lock()
            .await
            .changes_since(Some(&cursor))
            .await?;
        self.emit(
            "sync_progress",
            json!({"stage":"changes","total":updated.len()+deleted.len()}),
        );
        self.cache.upsert_reminders(&updated)?;
        if !updated.is_empty() {
            let ids = updated
                .iter()
                .map(|reminder| reminder.id.clone())
                .collect::<Vec<_>>();
            let tags = self.client.lock().await.tags_for_reminders(&ids).await?;
            for reminder in &updated {
                self.cache.replace_tags_for(
                    &reminder.id,
                    tags.get(&reminder.id).map(Vec::as_slice).unwrap_or(&[]),
                )?;
            }
        }
        for id in &deleted {
            let mut fields = serde_json::Map::new();
            fields.insert("deleted".into(), json!(1));
            self.cache.apply_local_edit(id, &fields)?;
            self.cache.clear_dirty(id, None)?;
        }
        if let Some(value) = new_cursor.as_deref() {
            self.cache.set_meta(CURSOR_KEY, Some(value))?;
        }
        self.cache
            .set_meta(LAST_SYNC_KEY, Some(&to_iso(Utc::now())))?;
        self.emit(
            "sync_finished",
            json!({"mode":"delta","updated":updated.len(),"deleted":deleted.len()}),
        );
        Ok(json!({"mode":"delta","updated":updated.len(),"deleted":deleted.len()}))
    }

    async fn flush_outbox(&self) -> Result<Value> {
        let Ok(_push_guard) = self.pushing.try_lock() else {
            return Ok(json!({"pushed":0,"conflicts":0,"failed":0,"skipped":true}));
        };
        let mut pushed = 0;
        let mut conflicts = 0;
        let mut failed = 0;
        for item in self.cache.pending(100)? {
            let result = match item.op.as_str() {
                "create" => self.client.lock().await.create(&item.payload).await,
                "update" => {
                    self.client
                        .lock()
                        .await
                        .update(&item.reminder_id, &item.payload, item.base_tag.as_deref())
                        .await
                }
                "delete" => {
                    self.client
                        .lock()
                        .await
                        .delete(&item.reminder_id, item.base_tag.as_deref())
                        .await
                }
                _ => {
                    self.cache.dequeue(item.seq)?;
                    continue;
                }
            };
            match result {
                Ok(remote) => {
                    if item.op == "create" {
                        self.cache.replace_id(&item.reminder_id, &remote.id)?;
                    }
                    self.cache
                        .clear_dirty(&remote.id, remote.change_tag.as_deref())?;
                    self.cache.upsert_reminders(&[remote])?;
                    self.cache.dequeue(item.seq)?;
                    pushed += 1;
                }
                Err(AppError::Conflict { detail, .. }) => {
                    let local = serde_json::to_value(
                        self.cache.reminder(&item.reminder_id)?.unwrap_or_default(),
                    )?;
                    let remote: Value =
                        serde_json::from_str(&detail).unwrap_or_else(|_| json!({"raw":detail}));
                    self.cache
                        .record_conflict(&item.reminder_id, &local, &remote)?;
                    if let Ok(reminder) = serde_json::from_value::<Reminder>(remote.clone()) {
                        self.cache
                            .clear_dirty(&item.reminder_id, reminder.change_tag.as_deref())?;
                        self.cache.upsert_reminders(&[reminder])?;
                    }
                    self.cache.dequeue(item.seq)?;
                    conflicts += 1;
                    self.emit("conflict", json!({"reminder_id":item.reminder_id}));
                }
                Err(error) => {
                    self.cache.record_failure(item.seq, &error.to_string())?;
                    failed += 1;
                    self.emit(
                        "push_failed",
                        json!({"reminder_id":item.reminder_id,"error":error.body()}),
                    );
                    if matches!(
                        error,
                        AppError::AuthRequired { .. }
                            | AppError::TwoFactorRequired { .. }
                            | AppError::TermsRequired { .. }
                    ) {
                        return Err(error);
                    }
                }
            }
        }
        Ok(json!({"pushed":pushed,"conflicts":conflicts,"failed":failed}))
    }
}
